File: week-5/lecture/7-radix-sort/radix-sort.cpp
/**
 * @file radix-sort.cpp
 * @brief C++ version of the week 5 radix-sort lecture example.
 */

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief Sort values by one digit position as part of radix sort.
 * @param array vector<int>& List of non-negative integers to reorder in place.
 * @param place int Current digit place, such as 1 for ones or 10 for tens.
 * @return void This function updates array in place and does not return a value.
 */
void countingSort(vector<int>& array, int place) {
    vector<int> output(array.size(), 0);
    vector<int> count(10, 0);

    for (int value : array) {
        ++count[(value / place) % 10];
    }
    for (size_t i = 1; i < count.size(); ++i) {
        count[i] += count[i - 1];
    }
    for (int i = static_cast<int>(array.size()) - 1; i >= 0; --i) {
        int digit = (array[i] / place) % 10;
        output[count[digit] - 1] = array[i];
        --count[digit];
    }
    array = output;
}

/**
 * @brief Sort non-negative integers with radix sort.
 * @param array vector<int>& List of non-negative integers to sort in place.
 * @return void This function updates array in place and does not return a value.
 */
void radixSort(vector<int>& array) {
    int maximum = *max_element(array.begin(), array.end());
    for (int place = 1; maximum / place > 0; place *= 10) {
        countingSort(array, place);
    }
}

/**
 * @brief Run the radix-sort lecture example.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> data = {121, 432, 564, 23, 1, 45, 788};
    radixSort(data);
    for (int value : data) {
        cout << value << ' ';
    }
    cout << '\n';
    return 0;
}

File: week-5/lecture/7-radix-sort/radix-sort.py
"""Radix Sort example.

This file belongs to the week 5 radix sort lecture example. It uses countingSort, radixSort to illustrate the main idea with a small runnable example.
"""

# Radix sort in Python


# Using counting sort to sort the elements in the basis of significant places
def countingSort(array, place):
    """Sort the values by one digit position as part of radix sort.

    Args:
        array (list): The list being processed by the algorithm.
        place (int): The place input used by this function.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    size = len(array)
    output = [0] * size
    count = [0] * 10

    # Calculate count of elements
    for i in range(0, size):
        index = array[i] // place
        count[index % 10] += 1

    # Calculate cumulative count
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Place the elements in sorted order
    i = size - 1
    while i >= 0:
        index = array[i] // place
        output[count[index % 10] - 1] = array[i]
        count[index % 10] -= 1
        i -= 1

    for i in range(0, size):
        array[i] = output[i]


# Main function to implement radix sort
def radixSort(array):
    # Get maximum element
    """Sort non-negative integers using radix sort.

    Args:
        array (list): The list being processed by the algorithm.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    max_element = max(array)

    # Apply counting sort to sort elements based on place value.
    place = 1
    while max_element // place > 0:
        countingSort(array, place)
        place *= 10


data = [121, 432, 564, 23, 1, 45, 788]
radixSort(data)
print(data)